orchagent/qosorch.cpp

```diff
@@ -1,10 +1,798 @@
+            // Item not found is not an error case
+            result = true;
+            break;
+    }
+    return result;
+}
+
+bool QosOrch::handlePortQosMapTable(Consumer& consumer)
+{
+    auto it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple tuple = it->second;
+    Port port;
+    string port_alias = kfvKey(tuple);
+    string op = kfvOp(tuple);
+    if(consumer.m_consumer->getTableName() != APP_PORT_QOS_MAP_TABLE_NAME) {
+        SWSS_LOG_ERROR("Key with invalid table type passed in %s, expected:%s\n", port_alias.c_str(), APP_PORT_QOS_MAP_TABLE_NAME);
+        return false;
+    }
+    if(!m_portsOrch->getPort(port_alias, port)) {
+        SWSS_LOG_ERROR("Port with alias:% not found\n", port_alias.c_str());
+        return false;
+    }
+
+    sai_attr_id_t attr_id;
+#ifdef MLNXSAI
+        attr_id = SAI_PORT_ATTR_QOS_DSCP_TO_TC_AND_COLOR_MAP;
+#else
+        attr_id = SAI_PORT_ATTR_QOS_DSCP_TO_TC_MAP;
+#endif
+    if(!tryApplyRefFieldToPort(op, dscp_to_tc_field_name, port, attr_id, &QosOrch::applyMapToPort, tuple)) {
+        SWSS_LOG_ERROR("Failed setting field:%s to port:%s, line:%d\n", dscp_to_tc_field_name.c_str(), port.m_alias.c_str(), __LINE__);
+        return false;
+    }
+    attr_id = SAI_PORT_ATTR_QOS_TC_TO_QUEUE_MAP;
+    if(!tryApplyRefFieldToPort(op, tc_to_queue_field_name, port, attr_id, &QosOrch::applyMapToPort, tuple)) {
+        SWSS_LOG_ERROR("Failed setting field:%s to port:%s, line:%d\n", tc_to_queue_field_name.c_str(), port.m_alias.c_str(), __LINE__);
+        return false;
+    }
+    return true;
+}
+
+void QosOrch::doTask(Consumer &consumer)
+{
+    if (consumer.m_toSync.empty()) {
+        return;
+    }
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end()) {
+        KeyOpFieldsValuesTuple tuple = it->second;
+        //
+        // make sure table is recognized, and we have handler for it
+        //
+        string qos_map_type_name = consumer.m_consumer->getTableName();
+        if(m_qos_type_maps.find(qos_map_type_name) == m_qos_type_maps.end()) {
+            SWSS_LOG_ERROR("Unrecognised qos table encountered:%s\n", qos_map_type_name.c_str());
+            //return;//?? [question.6]should be return or continue to retry later?
+            it++;
+            continue;
+        }
+        if(m_qos_handler_map.find(qos_map_type_name) == m_qos_handler_map.end()) {
+            SWSS_LOG_ERROR("No handler for key:%s found.\n", qos_map_type_name.c_str());
+            it++;
+            continue;
+        }
+        /*
+        [question.8] Decision on erasing an item from m_toSync to not loop forever.
+            When there is a failure in processing an item, need to decide - should leave it in m_toSync 
+            to retry later, or should remove from m_to_sync.
+
+            Returning bool may not be sufficient, may need to return different error codes
+            based on which to decide to do consumer.m_toSync.erase() or it++ to retry later.
+        */
```

Comment:
it seems that there're two options: retry or remove. so i'm wondering if boolean is enough?